PV Block , A/B partitions and AVB 2.0 support for Android/Xenvm

arminn opened the pull request.

arminn committed:
Handle verity disable command from Android

Android could update vbmeta during "adb disable-verity" command.
Do not fail due to changes in vbmeta if device is unlocked and
continue to boot using "orange" state.
https://source.android.com/security/verifiedboot/boot-flow#unlocked-devices

Reviewed-by: Volodymyr Babcuk <volodymyr_babchuk@epam.com>
Signed-off-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

arminn committed:
UPSTREAM: Introduce interface option to avb init

Using the interface option AVB could work on non-emmc
block devices(like pvblock, virtio and etc).

Reviewed-by: Volodymyr Babcuk <volodymyr_babchuk@epam.com>
Signed-off-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

arminn committed:
UPSTREAM: Introduce optional interface parameter to bcb

Interface parameter will give the ability to use bcb on
various block devices. The default logic won't be changed.

Reviewed-by: Volodymyr Babcuk <volodymyr_babchuk@epam.com>
Signed-off-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

arminn committed:
xen: pvblock: Introduce xenguest_arm64_android board

* This board configuration will be used for Android
  domain under Xen.
* Enable GPT support, since Android needs it.
* Change boot delay to 3 sec
* Add support of the bootm command
* Introduce bcb based logic
* Use ab_select to chose and support A/B partitions
* Enable AVB 2.0

Reviewed-by: Volodymyr Babcuk <volodymyr_babchuk@epam.com>
Signed-off-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

arminn committed:
xen: pvblock: Provide ability to include xenguest_arm64.h

Provide the ability to include and reuse xenguest_arm64.h
as a base configuration.

Reviewed-by: Volodymyr Babcuk <volodymyr_babchuk@epam.com>
Signed-off-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

andr2000 commented on arch/arm/Kconfig:
```diff
@@ -1725,6 +1725,16 @@ config TARGET_XENGUEST_ARM64
+
+config TARGET_XENGUEST_ARM64_ANDROID
+        bool "Xen guest ARM64 for Android"
+        select ARM64
```
this is common to all boards derived from XENGUEST_ARM64

andr2000 commented on board/xen/xenguest_arm64_android/Makefile:
```diff
@@ -0,0 +1,5 @@
+#
+# (C) Copyright 2020 EPAM Systems Inc.
+
+obj-y	:= xenguest_arm64.o
```
Do not copy the file, use the existing one: either have a symlink or obj-y	:= ../xenguest_arm64/xenguest_arm64.o

arminn committed:
xen: pvblock: Create a generic option for XENGUEST config

Create a generic option for XENGUEST config, which could
be used by derived xen guest modifications.

Reviewed-by: Volodymyr Babcuk <volodymyr_babchuk@epam.com>
Signed-off-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

andr2000 commented on cmd/avb.c:
```diff
@@ -228,6 +225,19 @@ int do_avb_get_uuid(struct cmd_tbl *cmdtp, int flag,
+static void do_avb_update_cmdline(bool unlocked, AvbSlotVerifyData *out_data)
+{
+	char *cmdline;
+	char *extra_args = avb_set_state(avb_ops, (unlocked) ? AVB_ORANGE : AVB_GREEN);
```
why brackets around unblocked?

andr2000 commented on cmd/avb.c:
```diff
@@ -228,6 +225,19 @@ int do_avb_get_uuid(struct cmd_tbl *cmdtp, int flag,
+	char *extra_args = avb_set_state(avb_ops, (unlocked) ? AVB_ORANGE : AVB_GREEN);
+	if (extra_args)
+		cmdline = append_cmd_line(out_data->cmdline,
+						extra_args);
```
alignment?

andr2000 commented on cmd/bcb.c:
```diff
@@ -48,6 +49,9 @@ static int bcb_is_misused(int argc, char *const argv[])
+			goto err;
+		break;
 	case BCB_CMD_FIELD_SET:
 		if (argc != 3)
```
argc != 4 - do we need this check here as well?

arminn replied:
yeah, we need to check both 3 & 4 old and new API

andr2000 commented on cmd/bcb.c:
```diff
@@ -23,6 +23,7 @@ enum bcb_cmd {
 static int bcb_dev = -1;
 static int bcb_part = -1;
 static struct bootloader_message bcb = { { 0 } };
+static struct blk_desc *bcb_blk_desc = NULL;
```
why do we need this global?

lorc commented:
"pvblock: android: Update boot params for r-car" confuses me.

First, you are add generic Xen android and then you are making it to look like RCAR. 
I think you need to squash this change to "xen: pvblock: Introduce xenguest_arm64_android board" and pretend that this is generic value :)

UPD: Actually, the same is true for the next couple of patches. Why you can't introduce Xen android platform with the correct config values in the first place?

lorc committed:
smcc: fix sign bit expansion

Signed ARM_SMCCC_FAST_CALL value is shifted the 31th bit. Then, it expanded to
64 bit value, which results in 1s in higher bits.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
optee: implement RPMB emulator

This emulator uses special RPMBEMUL partition on pvblock 0 device
to store RPMB state between reboots.

Enable it with CONFIG_OPTEE_RPMB_EMU=y

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
android_defconfig: enable RPMB emulator

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
rpmp_emul: fix global memory corruption

blk_dread will read blck_count * blk_size bytes which can be bigger than size of
struct rpmp_emu. Thus, it will overwrite any data that is stored past this
structure. In our particular case that lead to overwritten rpmb_part_offset
variable, which become 0. This, in turn, lead to GPT corrupt.

So we need to allocate memory buffer which is aligned with block size and
store RPMB emulator state there.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc commented on cmd/bcb.c:
```diff
@@ -320,16 +321,17 @@ static int do_bcb(struct cmd_tbl *cmdtp, int flag, int argc, char *const argv[])
 	bcb, CONFIG_SYS_MAXARGS, 1, do_bcb,
 	"Load/set/clear/test/dump/store Android BCB fields",
-	"load  <dev> <part>       - load  BCB from mmc <dev>:<part>\n"
+	"load  <dev> <part> [<interface>]      - load  BCB from <dev>:<part>[<interface>]\n"
```
looks like space before "[interface]" is missing

lorc commented on include/configs/xenguest_arm64_android.h:
```diff
@@ -0,0 +1,43 @@
+      fi;\0" \
+	"boot_pvblock=ab_select slot pvblock ${blk_device_id}#misc; run loadimage; fdt addr ${fdt_addr}; fdt get value bootargs /chosen bootargs; run update_bootargs; run avb_check; bootm ${boot_image_addr} ${boot_image_addr} ${fdt_addr};\0"\
+	"boot_pvblock_recovery=ab_select slot pvblock ${blk_device_id}#misc; run loadimage; fdt addr ${fdt_addr}; fdt get value bootargs /chosen bootargs; run update_bootargs_recovery; run avb_check; bootm ${boot_image_addr} ${boot_image_addr} ${fdt_addr};\0"\
+	"update_bootargs=set bootargs ${bootargs}\ androidboot.slot_suffix=_${slot} androidboot.force_normal_boot=1;\0" \
```
There are strange escaped space: "\ "

lorc commented on include/avb_verify.h:
```diff
@@ -98,4 +99,17 @@ static inline int get_boot_device(AvbOps *ops)
+{
+	struct AvbOpsData *data;
+
+	if (ops) {
```
you can simplify this to `if (ops && ops->user_data) return ops->userdata->blk`.
It is easier to read. Or can leave it as is

lorc reviewed (commented):
s/varity/verity in commit name
s/doue/due , s/ulocked/unlocked in the same commit description

lorc commented:
With above fixed: 
`Reviewed-by: Volodymyr Babcuk <volodymyr_babchuk@epam.com>`
For all patches except mine of course :)
